### ISSUE TITLE
rmlui build failure

### ISSUE THREAD

**Issue by BenKey** [microsoft/vcpkg#24413]
# Host Environment

- **OS:** Windows 10 Pro Version 21H2 build 19044.1645 with Windows Feature Experience Pack 120.2212.4170.0
- **Compiler:** Microsoft Visual Studio Community 2022 Version 17.1.5

# To Reproduce

`./vcpkg.exe install lua[core,cpp,tools] rmlui[core,freetype,lua] --triplet x64-windows`

# Failure logs

```
Starting package 10/10: rmlui:x64-windows
Building package rmlui[core,freetype,lua]:x64-windows...
-- Downloading https://github.com/mikke89/RmlUi/archive/4.3.tar.gz -> mikke89-RmlUi-4.3.tar.gz...
-- Extracting source D:/vcpkg/downloads/mikke89-RmlUi-4.3.tar.gz
-- Applying patch add-robin-hood.patch
-- Using source at D:/vcpkg/buildtrees/rmlui/src/4.3-7182cc7348.clean
-- Found external ninja('1.10.2').
-- Configuring x64-windows
-- Building x64-windows-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:158 (message):
    Command failed: "C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/bin/cmake.exe" --build . --config Debug --target install -- -v -j9
    Working Directory: D:/vcpkg/buildtrees/rmlui/x64-windows-dbg
    See logs for more information:
      D:\vcpkg\buildtrees\rmlui\install-x64-windows-dbg-out.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_build.cmake:111 (vcpkg_execute_build_process)
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_install.cmake:41 (vcpkg_cmake_build)
  ports/rmlui/portfile.cmake:27 (vcpkg_cmake_install)
  scripts/ports.cmake:145 (include)

error: building rmlui:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `git pull` and `.\vcpkg update`.
Then check for known issues at:
  https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+rmlui
You can submit a new issue at:
  https://github.com/microsoft/vcpkg/issues/new?template=report-package-build-failure.md&title=[rmlui]+Build+error
including:
  package: rmlui[core,freetype,lua]:x64-windows -> 4.3
    vcpkg-tool version: 2022-03-30-692785ac944e81417840c6de244fb3e18a4b35eb
    vcpkg-scripts version: 1ff261715 2022-04-25 (12 minutes ago)

Additionally, attach any relevant sections from the log files above.
```
[install-x64-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/8558713/install-x64-windows-dbg-out.log)

# Additional context

This error does not occur with the following command.

`./vcpkg.exe install rmlui[core,freetype,lua] --triplet x64-windows`

In other words, the problem is related to including 'lua[core,cpp,tools]' in the list of packages that is to be built.

**Comment by FrankXie05** [comment]
@BenKey Thanks for posting this issue, have you tried this command `./vcpkg.exe install lua[core,cpp,tools]:x64-windows`?

**Comment by BenKey** [comment]
@FrankXie05 The problem is not with building lua. The problem is with building rmlui. It occurs when building both lua and rmlui.

I cleaned my VCPKG directory. I ran the following command, and the build failed.

`./vcpkg.exe install lua[core,cpp,tools] rmlui[core,freetype,lua] --triplet x64-windows`

The command you specified does not build rmlui, it only builds lua.

**Comment by BenKey** [comment]
There is no need for more information. Use the specific command I used and you will see the error.

**Comment by FrankXie05** [comment]
@BenKey 
The port `lua` itself is built as a `c` library, the feature `cpp`  builds `lua` as a `c++` library, https://github.com/microsoft/vcpkg/blob/94c728df6b4c1c186e2ba58952f48ac744f4ef3b/ports/lua/CMakeLists.txt#L2
https://github.com/microsoft/vcpkg/blob/94c728df6b4c1c186e2ba58952f48ac744f4ef3b/ports/lua/vcpkg.json#L19
and the port `rmlui` dependency is regular `lua`, not depend on `lua[cpp]`. 
https://github.com/microsoft/vcpkg/blob/94c728df6b4c1c186e2ba58952f48ac744f4ef3b/ports/rmlui/vcpkg.json#L30
You can use the command` .\vcpkg.exe install lua[core,tools] rmlui[core,freetype,lua] --triplet x64-windows --recurse `,
Prerequisite: delete the local build cache before use.
````
PS E:\vcpkg> .\vcpkg.exe install lua[core,tools] rmlui[core,freetype,lua] --triplet x64-windows --recurse               
Computing installation plan...                                                                                          
The following packages will be built and installed:                                                                         
	lua[core,tools]:x64-windows -> 5.4.4#1                                                                                  
	rmlui[core,freetype,lua]:x64-windows -> 4.3                                                                         
Detecting compiler hash for triplet x64-windows...                                                                      
Restored 0 packages from C:\Users\test\AppData\Local\vcpkg\archives in 1.096 ms. Use --debug to see more details. Starting package 1/2: lua:x64-windows   
                                                                                
Building package lua[core,tools]:x64-windows...                                                                         
-- Using cached lua-5.4.4.tar.gz.                                                                                       
-- Extracting source E:/vcpkg/downloads/lua-5.4.4.tar.gz                                                                
-- Applying patch vs2015-impl-c99.patch                                                                                 
-- Applying patch fix-ios-system.patch                                                                                 
 -- Using source at E:/vcpkg/buildtrees/lua/src/lua-5-f0591007d5.clean                                                 
 -- Configuring x64-windows                                                                                            
 -- Building x64-windows-dbg                                                                                           
 -- Building x64-windows-rel                                                                                           
 -- Installing: E:/vcpkg/packages/lua_x64-windows/share/lua/usage                                                       
 -- Installing: E:/vcpkg/packages/lua_x64-windows/share/lua/copyright                                                  
 -- Performing post-build validation                                                                                   
 -- Performing post-build validation done                                                                              
 Stored binary cache: C:\Users\test\AppData\Local\vcpkg\archives\aa\aa3e48a0583ea00a943d67c7e2cb7ddcba06f940a6ab78ae1ca98cb95e12639a.zip    
 Installing package lua[core,tools]:x64-windows...                                                                     
 Elapsed time for package lua:x64-windows: 31.23 s                                                                    
 Starting package 2/2: rmlui:x64-windows                                                                              
 Building package rmlui[core,freetype,lua]:x64-windows...                                                              
 -- Using cached mikke89-RmlUi-4.3.tar.gz.                                                                             
 -- Cleaning sources at E:/vcpkg/buildtrees/rmlui/src/4.3-7182cc7348.clean. Use --editable to skip cleaning for the packages you specify.      
 -- Extracting source E:/vcpkg/downloads/mikke89-RmlUi-4.3.tar.gz                                                       
 -- Applying patch add-robin-hood.patch                                                                               
 -- Using source at E:/vcpkg/buildtrees/rmlui/src/4.3-7182cc7348.clean                                                 
 -- Configuring x64-windows                                                                                            
 -- Building x64-windows-dbg                                                                                           
 -- Building x64-windows-rel                                                                                           
 -- Installing: E:/vcpkg/packages/rmlui_x64-windows/share/rmlui/copyright                                              
 -- Performing post-build validation                                                                                   
 -- Performing post-build validation done                                                                              
 Stored binary cache: C:\Users\test\AppData\Local\vcpkg\archives\25\2501858603c81333ac66aaad137a94569aa12d82696ff759974eb3a3b84c05aa.zip    
 Installing package rmlui[core,freetype,lua]:x64-windows...                                                             
 Elapsed time for package rmlui:x64-windows: 1.994 min   
 
 Total elapsed time: 2.606 min  
 
 Use this package via the module FindLua that comes with CMake. To use in your CMakeLists.txt:                        
	 find_package(Lua REQUIRED)                                                                                           
	 target_include_directories(main PRIVATE ${LUA_INCLUDE_DIR})                                                          
	 target_link_libraries(main PRIVATE ${LUA_LIBRARIES}) 
	 
 The package rmlui provides CMake targets:   
 
	 find_package(RmlUi CONFIG REQUIRED)                                                                                  
	 target_link_libraries(main PRIVATE RmlLua RmlCore RmlDebugger)                                                       
PS E:\vcpkg> 
````

**Comment by JackBoosY** [comment]
```
FAILED: RmlLua.dll RmlLua.lib 
cmd.exe /C "cd . && "C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_dll --intdir=CMakeFiles\RmlLua.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\MIB055~1\2022\COMMUN~1\VC\Tools\MSVC\1431~1.311\bin\Hostx64\x64\link.exe  CMakeFiles\RmlLua.dir\Source\Lua\Colourb.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Colourf.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Context.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\ContextDocumentsProxy.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Document.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Element.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\ElementAttributesProxy.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\ElementChildNodesProxy.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\ElementInstancer.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\DataFormatter.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\DataSource.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementDataGrid.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementDataGridRow.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementForm.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementFormControl.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementFormControlDataSelect.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementFormControlInput.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementFormControlSelect.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementFormControlTextArea.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementTabSet.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\LuaDataFormatter.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\LuaDataSource.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\SelectOptionsProxy.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\ElementStyleProxy.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\ElementText.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Event.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\EventParametersProxy.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\GlobalLuaFunctions.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Interpreter.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Log.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Lua.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaDataModel.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaDocument.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaDocumentElementInstancer.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaElementInstancer.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaEventListener.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaEventListenerInstancer.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaPlugin.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaType.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\RmlUi.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\RmlUiContextsProxy.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Utilities.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Vector2f.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Vector2i.cpp.obj  /out:RmlLua.dll /implib:RmlLua.lib /pdb:RmlLua.pdb /dll /version:4.3 /machine:x64 /nologo    /debug /INCREMENTAL  RmlCore.lib  D:\vcpkg\installed\x64-windows\debug\lib\lua.lib  D:\vcpkg\installed\x64-windows\debug\lib\lua-c++.lib  D:\vcpkg\installed\x64-windows\debug\lib\freetyped.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib  && cd ."
LINK Pass 1: command "C:\PROGRA~1\MIB055~1\2022\COMMUN~1\VC\Tools\MSVC\1431~1.311\bin\Hostx64\x64\link.exe CMakeFiles\RmlLua.dir\Source\Lua\Colourb.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Colourf.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Context.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\ContextDocumentsProxy.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Document.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Element.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\ElementAttributesProxy.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\ElementChildNodesProxy.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\ElementInstancer.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\DataFormatter.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\DataSource.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementDataGrid.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementDataGridRow.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementForm.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementFormControl.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementFormControlDataSelect.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementFormControlInput.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementFormControlSelect.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementFormControlTextArea.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\ElementTabSet.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\LuaDataFormatter.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\LuaDataSource.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Elements\SelectOptionsProxy.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\ElementStyleProxy.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\ElementText.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Event.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\EventParametersProxy.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\GlobalLuaFunctions.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Interpreter.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Log.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Lua.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaDataModel.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaDocument.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaDocumentElementInstancer.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaElementInstancer.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaEventListener.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaEventListenerInstancer.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaPlugin.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\LuaType.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\RmlUi.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\RmlUiContextsProxy.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Utilities.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Vector2f.cpp.obj CMakeFiles\RmlLua.dir\Source\Lua\Vector2i.cpp.obj /out:RmlLua.dll /implib:RmlLua.lib /pdb:RmlLua.pdb /dll /version:4.3 /machine:x64 /nologo /debug /INCREMENTAL RmlCore.lib D:\vcpkg\installed\x64-windows\debug\lib\lua.lib D:\vcpkg\installed\x64-windows\debug\lib\lua-c++.lib D:\vcpkg\installed\x64-windows\debug\lib\freetyped.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:CMakeFiles\RmlLua.dir/intermediate.manifest CMakeFiles\RmlLua.dir/manifest.res" failed (exit code 1120) with the following output:
   Creating library RmlLua.lib and object RmlLua.exp
Vector2f.cpp.obj : error LNK2001: unresolved external symbol __imp_lua_gettop
Vector2i.cpp.obj : error LNK2001: unresolved external symbol __imp_lua_gettop
LuaPlugin.cpp.obj : error LNK2001: unresolved external symbol __imp_lua_gettop
```
Confirmed.

**Comment by JackBoosY** [comment]
Please test my PR https://github.com/microsoft/vcpkg/pull/24436.

Thanks.

**Comment by BenKey** [comment]
@JackBoosY I do not know how to test this change.

**Comment by JackBoosY** [comment]
@BenKey See `command line instructions` in my PR.

**Comment by BenKey** [comment]
@JackBoosY I have no idea what instructions you are referring to. I also do not know how to merge this pull reqest into my own fork.

**Comment by JackBoosY** [comment]
```
git clone https://github.com/JackBoosY/vcpkg.git
git checkout -b JackBoosY-dev/jack/24413 master
git pull https://github.com/JackBoosY/vcpkg.git dev/jack/24413
```

**Comment by BenKey** [comment]
@JackBoosY There is no change. I ran the commands you provided. I still get an error.

The log file is as follows: [install-x64-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/8673074/install-x64-windows-dbg-out.log).

**Comment by JackBoosY** [comment]
It's hard to believe that because I tested it locally.
Can you please provide a sample project to reproduce this issue?

Thanks.

**Comment by BenKey** [comment]
@JackBoosY I ran the commands you provided. I then ran bootstrap-vcpkg.bat. Finally, I ran the following command.

`./vcpkg.exe install lua[core,cpp,tools] rmlui[core,freetype,lua] --triplet x64-windows`

The errors occur while building rmlui.

**Comment by JackBoosY** [comment]
I still can't reproduce the issue using a fresh clone in my branch,
It looks fine when building the 2 ports.
Can you please provide the install log in _VCPKG_ROOT/buildtrees/lua_?

Thanks.

**Comment by BenKey** [comment]
@JackBoosY I think this is the log file you wanted.

[install-x64-windows-rel-out.log](https://github.com/microsoft/vcpkg/files/8691197/install-x64-windows-rel-out.log)

**Comment by JackBoosY** [comment]
Can you please update your vcpkg using `git pull`, remove port `lua` then try again?

Thanks.

**Comment by BenKey** [comment]
@JackBoosY I did that over the weekend. It works. Thanks.
